raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java

```diff
@@ -0,0 +1,132 @@
+        .getStateStore(TRACE_EMIT_TRIGGER_STORE);
+    this.groupingWindowTimeoutMs = ((Config) (context.appConfigs().get(RAW_SPANS_GROUPER_JOB_CONFIG)))
+        .getLong(SPAN_GROUPBY_SESSION_WINDOW_INTERVAL_CONFIG_KEY) * 1000;
+    this.outputTopic = To.child(OUTPUT_TOPIC_PRODUCER);
+    restorePunctuators();
+  }
+
+  @Override
+  public KeyValue<String, StructuredTrace> transform(String key, RawSpan value) {
+    ValueAndTimestamp<RawSpans> rawSpans = inflightTraceStore.get(key);
+    RawSpans agg = rawSpans != null ? rawSpans.value() : RawSpans.newBuilder().build();
+    // add the new span
+    agg.getRawSpans().add(value);
+
+    long currentTimeMs = System.currentTimeMillis();
+
+    if (logger.isDebugEnabled()) {
+      logger.debug("Updating ts=[{}] for trace_id=[{}]", Instant.ofEpochMilli(currentTimeMs), key);
+    }
+    // store the current system time as the last update timestamp
+    inflightTraceStore.put(key, ValueAndTimestamp.make(agg, currentTimeMs));
+
+    /*
+     check if any prior trace emit ts was stored. It is stored when an entry for a trace doesn't exist.
+     See the {@link TraceEmitPunctuator} for details on how the traceEmitTs is updated
+    */
+    Long triggerTs = traceEmitTriggerStore.get(key);
+    boolean firstEntry = (triggerTs == null);
+    /*
+     the trace emit ts is essentially currentTs + sessionTimeoutMs
+     i.e. if there is no span added in the next 'sessionTimeoutMs' interval
+     then the trace can be finalized and emitted
+    */
+
+    long traceEmitTs = currentTimeMs + groupingWindowTimeoutMs;
+    if (logger.isDebugEnabled()) {
+      logger.debug("Updating trigger_ts=[{}] for trace_id=[{}]", Instant.ofEpochMilli(traceEmitTs),
+          key);
+    }
+    traceEmitTriggerStore.put(key, traceEmitTs);
+
+    // if no prior emitTs is found then add it.
+    if (firstEntry) {
+      schedulePunctuator(key);
+    }
+    // the punctuator will emit the trace
+    return null;
+  }
+
+  private void schedulePunctuator(String key) {
+    TraceEmitPunctuator punctuator = new TraceEmitPunctuator(key, context, inflightTraceStore,
+        traceEmitTriggerStore,
+        outputTopic, groupingWindowTimeoutMs);
+    Cancellable cancellable = context
+        .schedule(Duration.ofMillis(groupingWindowTimeoutMs), PunctuationType.WALL_CLOCK_TIME,
+            punctuator);
+    punctuator.setCancellable(cancellable);
+    logger.debug("Scheduled a punctuator to emit trace for trace_id=[{}] to run after [{}] ms", key,
+        groupingWindowTimeoutMs);
+  }
+
+  @Override
+  public void close() {
+  }
+
+  /**
+   * Punctuators are not persisted - so on restart we recover punctuators and schedule them to run
+   * after {@link RawSpansGroupingTransformer#groupingWindowTimeoutMs}
+   */
+  void restorePunctuators() {
+    long count = 0;
+    Instant start = Instant.now();
+    try (KeyValueIterator<String, Long> it = traceEmitTriggerStore.all()) {
```

Comment:
Hmm, could this be costly - `traceEmitTriggerStore.all` ocationally?

Reply:
hopefully this store is not very huge...

Reply:
should a comment be made about what happens if one key is run twice? ex is there risk of this? If there isn't why? If there is, why does it not matter?